trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/library/LibraryModelHandler.java

```diff
@@ -0,0 +1,167 @@
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.container.c.ICType;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.exception.UnsupportedSyntaxException;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.Result;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+import de.uni_freiburg.informatik.ultimate.core.model.services.ILogger;
+
+/**
+ * The {@link LibraryModelHandler} creates the translation for various functions and types where we have our own
+ * specification or implementation. This is typically the case for functions defined in the C standard, but also for
+ * various standard libraries or SV-COMP extensions.
+ *
+ * @author Markus Lindenmann,
+ * @author Matthias Heizmann
+ * @author Daniel Dietsch (dietsch@informatik.uni-freiburg.de)
+ * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
+ */
+public class LibraryModelHandler {
+	private final LocationFactory mLocationFactory;
+	private final Map<String, IFunctionModelHandler> mFunctionModels;
+	private final Map<String, ICType> mTypeModels;
+	private final Map<String, IASTNode> mFunctionTable;
+	private final FlatSymbolTable mSymboltable;
+	private final TranslationSettings mSettings;
+	private final ILogger mLogger;
+
+	public LibraryModelHandler(final ILogger logger, final Map<String, IASTNode> functionTable,
+			final FlatSymbolTable symboltable, final TranslationSettings settings,
+			final LocationFactory locationFactory, final List<ILibraryModel> libraryModels) {
+		mLogger = logger;
+		mFunctionTable = functionTable;
+		mSymboltable = symboltable;
+		mSettings = settings;
+		mLocationFactory = locationFactory;
+		mFunctionModels = getFunctionModels(libraryModels);
+		mTypeModels = getTypeModels(libraryModels);
+	}
+
+	/**
+	 * Check if the given function has an "integrated" specification or implementation and return a {@link Result} that
+	 * contains a translation of the function if this is the case.
+	 *
+	 * Return null otherwise.
+	 *
+	 * @param main
+	 * @param node
+	 * @param astIdExpression
+	 * @return
+	 */
+	public Result translateStandardFunction(final IDispatcher main, final IASTFunctionCallExpression node,
+			final IASTIdExpression astIdExpression) {
+		assert node.getFunctionNameExpression() == astIdExpression
+				: "astIdExpression is not the name of the called function";
+		final String name = astIdExpression.getName().toString();
+
+		final IFunctionModelHandler functionModel = mFunctionModels.get(name);
+		if (functionModel != null) {
+			final String transformedName = mSymboltable.applyMultiparseRenaming(node.getContainingFilename(), name);
+			final IASTNode funDecl = mFunctionTable.get(transformedName);
+			if (funDecl instanceof IASTFunctionDefinition) {
+				// it is a function that already has a body
+				if (!mSettings.checkErrorFunction() || !"reach_error".equals(transformedName)) {

```

Comment:
Could `reach_error` not also be handled by `SvcompLibraryModel`?

Reply:
It is handled by `SvcompLibraryModel`. This code here is just responsible that a warning is shown if `reach_error` is defined (as done in many tasks of SV-COMP). It was just moved from `StandardFunctionHandler`, but I might be useful to always show this warning, not just for `reach_error`.

Reply:
Ah, that is risky. I'm sceptical that it is a good idea to override functions that are actually defined in the C code.
- This may for now be ok for standard library functions.
- But it already gets somewhat critical with SV-COMP functions -- redefining them is certainly not undefined. Thus the warning for `reach_error`.
- E.g. for linux functions it may depend on the verified program. If we are analysing some kind of kernel implementation, we should definitely not substitute parts of it with our model.

Reply:
I agree, this was not done fully on purpose. Therefore I reverted this change and just clarified the existing code insteadb(9030a5b).